internal/framework/generic_resource.go

```diff
@@ -333,46 +364,167 @@ func isNetworkError(err error) bool {
+						map[string]any{"event": event.Type, "data": event.Data})
+					return json.Unmarshal([]byte(event.Data), state)
+				})
+				if err == nil {
+					sendToChannel(ctx, stream.eventChannel, true)
+				} else {
+					sendToChannel(ctx, stream.errChannel, err)
+				}
+			}
+		})
+		// If context is not done, there was an error consuming SSE messages
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			tflog.Info(ctx, "Downgrading from SSE to polling", map[string]any{"error": err})
+		}
+		// Use polling to generate resource events
+		for {
+			err := stream.withLock(func() error {
+				return state.Read(ctx, r.client.Client)
+			})
+			if err == nil {
+				sendToChannel(ctx, stream.eventChannel, true)
+			} else if helper.IsNotFound(err) {
+				sendToChannel(ctx, stream.eventChannel, false)
+			} else if isNetworkError(err) {
+				// Suppress network errors, which may be transient and retriable
+				tflog.Info(ctx, "Suppressing network error while awaiting readiness",
+					map[string]any{"resourceType": r.TypeName, "error": err.Error()})
+			} else {
+				sendToChannel(ctx, stream.errChannel, err)
+			}
+			// Wait for polling interval or until context is done
+			select {
+			case <-time.After(POLLING_INTERVAL):
+				continue
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return &stream
+}
+
 func (r *GenericResource) AwaitReady(ctx context.Context, state ResourceState, operation string) error {
 	timeout := r.GetTimeout(operation, READINESS_TIMEOUT)
 	if timeout == 0 {
 		tflog.Info(ctx, "Not waiting for resource to achieve desired state because timeout is 0",
 			map[string]any{"resourceType": r.TypeName, "operation": operation})
 		return nil
 	}
-	consecutiveFailed := 0
 	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
+	stream := r.stream(ctx, state)
+	defer func() {
+		cancel()
+		stream.wg.Wait()
+	}()
+	var readyErr error
+	var failedSince time.Time
 	for {
+		var err error
+		done := false
+		select {
+		case exists := <-stream.eventChannel:
+			// Check that resource still exists
+			if !exists {
+				return errors.New("Resource no longer exists")
+			}
+		case <-time.After(FAILURE_THRESHOLD):
+			// Check readiness periodically even if not triggered by channel
+		case err = <-stream.errChannel:
+			// Check error encountered on channel
+		case <-ctx.Done():
+			// Check that timeout has not expired
+			done = true
+		}
+		// If error was encountered on channel or timeout expired, return error
+		if err != nil || done {
+			if done || os.IsTimeout(err) && ctx.Err() == context.DeadlineExceeded {
+				if readyErr != nil {
+					return fmt.Errorf("Timed out after %s: %s", timeout, readyErr.Error())
+				} else {
+					return fmt.Errorf("Timed out after %s", timeout)
+				}
+			}
+			return err
+		}
 		// Check if resource is ready
-		readyErr := state.CheckReady(ctx, r.client.Client)
+		readyErr = stream.withLock(func() error {
+			return state.CheckReady(ctx, r.client.Client)
```

Comment:
Compared to status polling, the SSE stream will generate a lot more events as now the consumer will trigger on every change of the DBaaS resource. This may cause problems with `DatabaseResourceModel.CheckReady()` method which performs a REST API call to `UpdateDbaPassword`.

Does it make sense for the caller to provide an event predicate so that it's notified only on specific cases?

Reply:
Good point about the database readiness check performing `POST /databases/.../dbaPassword`. But that is only done if the database is in ready state, and we do want to verify the DBA password in that case.

Reply:
Ah, you are correct. I overlooked the `CheckReady()` implementation.